server/migrations/v2.13.5-series-column-unique.js

```diff
@@ -0,0 +1,126 @@
+const Logger = require('../Logger')
+
+/**
+ * @typedef MigrationContext
+ * @property {import('sequelize').QueryInterface} queryInterface - a suquelize QueryInterface object.
+ * @property {import('../Logger')} logger - a Logger object.
+ *
+ * @typedef MigrationOptions
+ * @property {MigrationContext} context - an object containing the migration context.
+ */
+
+/**
+ * This upward migration script cleans any duplicate series in the `Series` table and
+ * adds a unique index on the `name` and `libraryId` columns.
+ *
+ * @param {MigrationOptions} options - an object containing the migration context.
+ * @returns {Promise<void>} - A promise that resolves when the migration is complete.
+ */
+async function up({ context: { queryInterface, logger } }) {
+  // Upwards migration script
+  logger.info('UPGRADE BEGIN: 2.13.5-series-column-unique ')
+
+  // Use the queryInterface to get the series table and find duplicates in the `name` column
+  const [duplicates] = await queryInterface.sequelize.query(`
+    SELECT name, libraryId, MAX(updatedAt) AS latestUpdatedAt, COUNT(name) AS count
+    FROM Series
+    GROUP BY name, libraryId
+    HAVING COUNT(name) > 1
+  `)
+
+  // Print out how many duplicates were found
+  logger.info(`[2.13.5 migration] Found ${duplicates.length} duplicate series`)
+
+  // Iterate over each duplicate series
+  for (const duplicate of duplicates) {
+    // Report the series name that is being deleted
+    logger.info(`[2.13.5 migration] Deduplicating series "${duplicate.name}" in library ${duplicate.libraryId}`)
+
+    // Get all the most recent series which matches the `name` and `libraryId`
+    const [mostRecentSeries] = await queryInterface.sequelize.query(
+      `
+        SELECT id
+        FROM Series
+        WHERE name = :name AND libraryId = :libraryId
+        ORDER BY updatedAt DESC
+        LIMIT 1
+        `,
+      {
+        replacements: {
+          name: duplicate.name,
+          libraryId: duplicate.libraryId
+        },
+        type: queryInterface.sequelize.QueryTypes.SELECT
+      }
+    )
+
+    if (mostRecentSeries) {
+      // Update all BookSeries records for this series to point to the most recent series
+      const [seriesUpdated] = await queryInterface.sequelize.query(
+        `
+          UPDATE BookSeries
+          SET seriesId = :mostRecentSeriesId
+          WHERE seriesId IN (
+            SELECT id
+            FROM Series
+            WHERE name = :name AND libraryId = :libraryId
+            AND id != :mostRecentSeriesId
+          )
+          `,
+        {
+          replacements: {
+            name: duplicate.name,
+            libraryId: duplicate.libraryId,
+            mostRecentSeriesId: mostRecentSeries.id
+          }
+        }
+      )
```

Comment:
I believe there's an issue here that the test doesn't uncover because the bookSeries table you create there is not set-up correctly.

If you examine an actual ABS database, you will see that the bookSeries table has a unique contraint on [bookId, seriesId]. 

This means that if some book with id `book1` points to two series duplicates with ids `seriesDup1` and `seriesDup2`, then before this query there are two bookSeries records `[book1, seriesDup1]` and `[book1, seriesDup2]`, but after this query there will be `[book1, seriesDup2]` and `[book1, seriesDup2]`, which violates the unique constraint and will throw an exception.

I don't know if such a situation occurs in user databases, but we have to assume it does.

Your test won't uncover this (even if it has the right test) since it creates the bookSeries table without the unique constraint.

Reply:
I'm not seeing the unique constraint on [bookId, seriesId]
```
CREATE TABLE bookSeries (
    id        UUID,
    sequence  VARCHAR (255),
    bookId    UUID,
    seriesId  UUID,
    createdAt DATETIME,
    CONSTRAINT BOOKSERIES_PK PRIMARY KEY (
        id
    ),
    CONSTRAINT FK_bookSeries_books FOREIGN KEY (
        bookId
    )
    REFERENCES books (id) ON DELETE CASCADE
                          ON UPDATE CASCADE,
    CONSTRAINT FK_bookSeries_series_2 FOREIGN KEY (
        seriesId
    )
    REFERENCES series (id) ON DELETE CASCADE
                           ON UPDATE CASCADE
);

```

Reply:
Yeah, that's what I was basing my tables off of. If you look at the actual SQLite database, you can see the table is created as below (the unique constraint at the very end of the line)
```
CREATE TABLE `bookSeries` (`id` UUID PRIMARY KEY, `sequence` VARCHAR(255), `bookId` UUID REFERENCES `books` (`id`) ON DELETE CASCADE ON UPDATE CASCADE, `seriesId` UUID REFERENCES `series` (`id`) ON DELETE CASCADE ON UPDATE CASCADE, `createdAt` DATETIME, UNIQUE (`bookId`, `seriesId`))
```

Reply:
@advplyr where is the statement above coming from?

Reply:
Probably here? https://github.com/advplyr/audiobookshelf/blob/b35fabbe55146f7d2a721a4bbfea51cb3f5c0ad4/server/models/BookSeries.js#L32-L66

Reply:
Sure, this is the Sequelize init for the BookSeries model. I was asking where the SQL code was taken from.

These two lines
```js
     book.belongsToMany(series, { through: BookSeries }) 
     series.belongsToMany(book, { through: BookSeries }) 
```
are responsible for the implicit creation of the unique constraint on [bookId, seriesId].
This is explained [here](https://sequelize.org/docs/v6/advanced-association-concepts/advanced-many-to-many/). Specifically, it says "_... A composite unique key will be established on these two columns._"

Reply:
I installed SQLiteStudio after you mentioned it (previously used DBeaver) and the DDL tab is where I got that from.
I just switched over to my new PC and currently working on getting my dev environment setup so don't have those installed atm.

Reply:
You're right. My new db has the UNIQUE constraint.

Reply:
Ah, that in itself indicates that databases started at different points in time are not consistent (sigh)...
This requires another migration...

Reply:
Does the migration manager support multiple migration scripts for the same release?

Reply:
No, mutliple migrations scripts in the same version aren't supported. 

To be clear, I am not requiring any additional changes in your migration. I was just saying that fact that we have the existing codebase running databases with different schemas is worrying, but wasn't suggesting that we do anything immediate about it.

Reply:
I'm not sure how that could have happened since that relationship has always been setup like that since the first release of that table. It could be that I'm the only one with that difference and it happened during the initial development.

Reply:
I see, thanks!